File: web/src/types/task.ts
// Mirrors the taskd backend DTOs (pkg/dto). Field names are snake_case to match
// the JSON tags produced by the Go side.
export type Protocol = "http" | "https" | "grpc";

export type Status =
  | "scheduled"
  | "claimed"
  | "running"
  | "succeeded"
  | "failed"
  | "dead"
  | "canceled";

export type ExecutionResult = "success" | "failure";

export type Task = {
  id: string;
  name: string;
  ref_id: string;
  protocol: Protocol;
  address: string;
  params: Record<string, unknown>;
  exec_time: string | null;
  status: Status;
  attempts: number;
  max_retries: number;
  last_error: string;
  locked_until: string | null;
  cron: string;
};

export type TaskRecord = {
  id: string;
  task_id: string;
  attempt: number;
  result: ExecutionResult;
  protocol: Protocol;
  instance_id: string;
  error_message: string;
  started_at: string;
  finished_at: string;
  duration_ms: number;
  response: unknown;
  created_at: string;
};

export type ListTasksResp = {
  tasks: Task[];
  total: number;
  limit: number;
  offset: number;
};

export type ListTaskRecordsResp = {
  records: TaskRecord[];
  limit: number;
  offset: number;
};

export type CreateTaskBody = {
  name: string;
  ref_id: string;
  protocol: Protocol;
  address: string;
  params: Record<string, unknown>;
  exec_time: number; // unix seconds
  max_retries: number;
  cron?: string;
};

export type CreateTaskReq = { body: CreateTaskBody };

File: pkg/dto/task.go
package dto

import (
	"time"

	"github.com/flametest/taskd/internal/constant/enum"
)

// Task is the API representation of a task. Unlike domain.Task it exposes Status
// (the domain keeps it private) and represents LastError as a string, so the
// JSON response includes the task's current state.
type Task struct {
	Id          string                 `json:"id"`
	Name        string                 `json:"name"`
	RefId       string                 `json:"ref_id"`
	Protocol    enum.Protocol          `json:"protocol"`
	Address     string                 `json:"address"`
	Params      map[string]interface{} `json:"params"`
	ExecTime    *time.Time             `json:"exec_time"`
	Status      enum.Status            `json:"status"`
	Attempts    int                    `json:"attempts"`
	MaxRetries  int                    `json:"max_retries"`
	LastError   string                 `json:"last_error"`
	LockedUntil *time.Time             `json:"locked_until"`
	Cron        string                 `json:"cron"`
	CreatedAt   time.Time              `json:"created_at"`
}

// ListTasksResp is the response of GET /v1/tasks.
type ListTasksResp struct {
	Tasks  []*Task `json:"tasks"`
	Total  int64   `json:"total"`
	Limit  int     `json:"limit"`
	Offset int     `json:"offset"`
}
